### PR TITLE
ConstantFold logl calls

### PR DESCRIPTION
**Pull request by MDevereau**
This is a follow up patch from #90611 which folds logl calls in the same manner as log.f128 calls. logl suffers from the same problem as logf128 of having slow calls to fp128 log functions which can be constant folded. However, logl is emitted with -fmath-errno and log.f128 is emitted by -fno-math-errno by certain intrinsics.

### CODE REVIEW

**Comment by arsenm** [comment]
Description is imprecise, intrinsic or libcall depends on using -fmath-errno/-fno-math-errno only. Don't mention -Ofast

**Review by davemgreen — COMMENTED** [review]
Could you add tests with out-of-range values? 0 and inf and nan and negatives?

**Comment by MDevereau** [comment]
> Could you add tests with out-of-range values? 0 and inf and nan and negatives?

Done, I've removed the zero and negative checks as well to keep it consistent with the logf.128 behaviour.

**Comment by davemgreen** [comment]
> Done, I've removed the zero and negative checks as well to keep it consistent with the logf.128 behaviour.

It needs to produce the same errno and fpexeption values. I can see that the intrinsic remains, but would recommend making it work the same as floats: https://godbolt.org/z/T5x7foe4Y

File: llvm/lib/Analysis/ConstantFolding.cpp [hunk 1]
```diff
@@ -2085,15 +2085,18 @@ static Constant *ConstantFoldScalarCall1(StringRef Name,
...
-      }
+      const APFloat &Fp128APF = Op->getValueAPF();
+      if (IntrinsicID == Intrinsic::log)
+        return ConstantFP::get(Ty, logf128(Fp128APF.convertToQuad()));
     }
+
+    LibFunc Fp128Func = NotLibFunc;
+    if (Ty->isFP128Ty() && TLI->getLibFunc(Name, Fp128Func) &&
+        TLI->has(Fp128Func) && Fp128Func == LibFunc_logl)
+      return ConstantFP::get(Ty, logf128(Op->getValueAPF().convertToQuad()));
```
Comment (arsenm) [inline llvm/lib/Analysis/ConstantFolding.cpp:2099]: Combine these under 1 isFP128Ty check?

**Comment by MDevereau** [comment]
> > Done, I've removed the zero and negative checks as well to keep it consistent with the logf.128 behaviour.
> 
> It needs to produce the same errno and fpexeption values. I can see that the intrinsic remains, but would recommend making it work the same as floats: https://godbolt.org/z/T5x7foe4Y

I think I've done what you asked for. I'm a bit unsure about the infinity test case though.

File: llvm/lib/Analysis/ConstantFolding.cpp (hunk 1 above)
Comment (MDevereau) [reply llvm/lib/Analysis/ConstantFolding.cpp]: Done

**Review by davemgreen — COMMENTED** [review]
As far as I understand the difference between logl and llvm.log.f128 is that logl needs to set errno in the right places and has side-effects, llvm.log.f128 doesn't require that. That is why we convert logl to llvm.log.f128 under Ofast/-fno-math-errno. So `llvm.log.f128(0)` should still be fine to be constant-folded to Nan, it is only the logl calls that need to be more careful.

**Comment by MDevereau** [comment]
> As far as I understand the difference between logl and llvm.log.f128 is that logl needs to set errno in the right places and has side-effects, llvm.log.f128 doesn't require that. That is why we convert logl to llvm.log.f128 under Ofast/-fno-math-errno. So `llvm.log.f128(0)` should still be fine to be constant-folded to Nan, it is only the logl calls that need to be more careful.

Makes sense, I've removed the checks for log.f128 then. With logl infinity, I was unsure whether the fact it hasn't retained the original logl call was correct behaviour or not.

**Review by davemgreen — APPROVED** [review]
> Makes sense, I've removed the checks for log.f128 then. With logl infinity, I was unsure whether the fact it hasn't retained the original logl call was correct behaviour or not.

It looks like it doesn't generate a pole error for infinity when I tried it (only for zero), so I think that sounds OK. From what I can tell this LGTM, thanks.

**Pull request merged**